sel/src/Sel/SecretKey/AuthenticatedEncryption.hs

```diff
@@ -0,0 +1,208 @@
+  (Nonce hk1) == (Nonce hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrEq hk1 hk2 cryptoSecretboxKeyBytes
+
+-- |
+--
+-- @since 0.0.1.0
+instance Ord Nonce where
+  compare (Nonce hk1) (Nonce hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrOrd hk1 hk2 cryptoSecretboxKeyBytes
+
+-- | Generate a new random secret key.
+--
+-- @since 0.0.1.0
+newSecretKey :: IO SecretKey
+newSecretKey = do
+  fPtr <- Foreign.mallocForeignPtrBytes (fromIntegral cryptoSecretboxKeyBytes)
+  Foreign.withForeignPtr fPtr $ \ptr ->
+    cryptoSecretboxKeygen ptr
+  pure $ SecretKey fPtr
+
+-- | Generate a new random nonce.
+-- Only use it once per exchanged message.
+--
+-- @since 0.0.1.0
+newNonce :: IO Nonce
+newNonce = do
+  (fPtr :: ForeignPtr CUChar) <- Foreign.mallocForeignPtrBytes (fromIntegral cryptoSecretboxNonceBytes)
+  Foreign.withForeignPtr fPtr $ \ptr ->
+    randombytesBuf (Foreign.castPtr @CUChar @Word8 ptr) cryptoSecretboxNonceBytes
+  pure $ Nonce fPtr
+
+-- | A ciphertext consisting of an encrypted message and an authentication tag.
+--
+-- @since 0.0.1.0
+data Hash = Hash
+  { messageLength :: CULLong
+  , hashForeignPtr :: ForeignPtr CUChar
+  }
+
+-- |
+--
+-- @since 0.0.1.0
+instance Eq Hash where
+  (Hash messageLength1 hk1) == (Hash messageLength2 hk2) =
+    unsafeDupablePerformIO $ do
+      result1 <- foreignPtrEq hk1 hk2 (fromIntegral messageLength1 + cryptoSecretboxMACBytes)
+      pure $ (messageLength1 == messageLength2) && result1
+
+-- |
+--
+-- @since 0.0.1.0
+instance Ord Hash where
+  compare (Hash messageLength1 hk1) (Hash messageLength2 hk2) =
+    unsafeDupablePerformIO $ do
+      result1 <- foreignPtrOrd hk1 hk2 (fromIntegral messageLength1 + cryptoSecretboxMACBytes)
+      pure $ compare messageLength1 messageLength2 <> result1
+
+-- | Create an authenticated hash from a message, a secret key that must remain secret, and a one-time cryptographic nonce
+-- that must never be re-used with the same secret key to encrypt another message.
+--
+-- @since 0.0.1.0
+encrypt
+  :: StrictByteString
+  -- ^ Message to encrypt.
+  -> SecretKey
+  -- ^ Secret key generated with 'newSecretKey'.
+  -> Nonce
+  -- ^ One-time use number generated with 'newNonce'.
+  -> IO Hash
+encrypt message (SecretKey secretKeyForeignPtr) (Nonce nonceForeignPtr) =
+  BS.unsafeUseAsCStringLen message $ \(cString, cStringLen) -> do
+    hashForeignPtr <- Foreign.mallocForeignPtrBytes (cStringLen + fromIntegral cryptoSecretboxMACBytes)
+    Foreign.withForeignPtr hashForeignPtr $ \hashPtr ->
+      Foreign.withForeignPtr secretKeyForeignPtr $ \secretKeyPtr ->
+        Foreign.withForeignPtr nonceForeignPtr $ \noncePtr -> do
+          void $
+            cryptoSecretboxEasy
+              hashPtr
+              (Foreign.castPtr @CChar @CUChar cString)
+              (fromIntegral @Int @CULLong cStringLen)
+              noncePtr
+              secretKeyPtr
+    pure $ Hash (fromIntegral @Int @CULLong cStringLen) hashForeignPtr
```

Comment:
Since nonces are meant to be single-use, it would be better to have:

``encrypt :: StrictByteString -> SecretKey -> IO (Nonce, Hash)``

That way, nonces can never be re-used.

Reply:
Yes, very smart. :)